File: ets2panda/migrator/test/ts/function_expression.ts
/*
 * Copyright (c) 2022-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */ 

type MyFunction = (x: number, y: boolean) => string; 

interface I {
    m(): number;
}

function createFunc(): () => number {
    return function () { return 100; };
}

function filter(array: number[], predicate: (x: number) => boolean): number[] {
    let newArray: number[] = [];
    for (let element of array) {
        if (predicate(element))
            newArray.push(element);
    }
    return newArray;
}

function main(): void {
    let empty = function() { };
    empty();

    let multiply = function(x: number, y: number): number { return x * y; }
    console.log(multiply(5, 10));

    let foo = createFunc();
    let a: number = foo();
    let b: number = createFunc()();

    // Infer parameter's and return type.
    let bar: MyFunction = function(x, y) { return "value"; };
    console.log(bar(10, true));

    let restParam = (x: number, y: number, ...rest: number[]) => { 
        for (let z of rest) {
            console.log(x + y + z);
        }
    }
    restParam(1, 2);
    restParam(1, 2, 3);
    restParam(1, 2, 3, 4);
    restParam(1, 2, 3, 4, 5);

    // Immediately invoked function expression
    let foobar = function() { return "get result immediately";}();

    (function () {
        console.log("foo!");
    })();
      
    void function () {
        console.log("bar!");
    }();

    !function () { return false; }();
    
    +function () { return 0; }();

    // Named function expression with recursive call.
    let factorial = function func(n: number): number {
        if (n === 1)
            return 1;
        else
            return n * func(n - 1);
    }
    console.log(factorial(5));
    
    // Function as a callback.
    let array = [1, 2, 3, 4, 5, 6];
    let double = array.map(function(e) { return e * 2; });
    let even = filter(array, function(x) { return x % 2 === 0; });
    let isOdd = function(x: number) { return x % 2 !== 0; };
    let odd = filter(array, isOdd);

    // Generic function expression.
    let generic1 = function <T, E> (t: T, e: E) { return t; };
    generic1(100, "apple");
    let baz: string = generic1<string, string>("arg1", "arg2");

    let generic2 = function <W extends I> (w: W): void {
        w.m();
    }
    generic2({ m() { return 10; } });
}

File: ets2panda/ir/expressions/functionExpression.cpp
/**
 * Copyright (c) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "functionExpression.h"

#include "compiler/core/pandagen.h"
#include "compiler/core/ETSGen.h"
#include "checker/TSchecker.h"
#include "ir/astDump.h"
#include "ir/base/scriptFunction.h"
#include "ir/expressions/identifier.h"
#include "ir/statements/variableDeclarator.h"

namespace panda::es2panda::ir {
void FunctionExpression::Iterate(const NodeTraverser &cb) const
{
    cb(func_);
}

void FunctionExpression::Dump(ir::AstDumper *dumper) const
{
    dumper->Add({{"type", "FunctionExpression"}, {"function", func_}});
}

void FunctionExpression::Compile(compiler::PandaGen *pg) const
{
    pg->DefineFunction(func_, func_, func_->Scope()->InternalName());
}

void FunctionExpression::Compile([[maybe_unused]] compiler::ETSGen *etsg) const
{
    UNREACHABLE();
}

checker::Type *FunctionExpression::Check(checker::TSChecker *checker)
{
    binder::Variable *func_var = nullptr;

    if (func_->Parent()->Parent() != nullptr && func_->Parent()->Parent()->IsVariableDeclarator() &&
        func_->Parent()->Parent()->AsVariableDeclarator()->Id()->IsIdentifier()) {
        func_var = func_->Parent()->Parent()->AsVariableDeclarator()->Id()->AsIdentifier()->Variable();
    }

    checker::ScopeContext scope_ctx(checker, func_->Scope());

    auto *signature_info = checker->Allocator()->New<checker::SignatureInfo>(checker->Allocator());
    checker->CheckFunctionParameterDeclarations(func_->Params(), signature_info);

    auto *signature =
        checker->Allocator()->New<checker::Signature>(signature_info, checker->GlobalResolvingReturnType(), func_);
    checker::Type *func_type = checker->CreateFunctionTypeWithSignature(signature);

    if (func_var != nullptr && func_var->TsType() == nullptr) {
        func_var->SetTsType(func_type);
    }

    signature->SetReturnType(checker->HandleFunctionReturn(func_));

    func_->Body()->Check(checker);

    return func_type;
}

checker::Type *FunctionExpression::Check([[maybe_unused]] checker::ETSChecker *checker)
{
    return nullptr;
}
}  // namespace panda::es2panda::ir
